File: backend/src/services/auction.service.ts
import { prisma } from '../lib/prisma.js';

/**
 * Evaluate whether a bid should trigger an auction time extension.
 * Implements the British Auction extension rules:
 * - ANY_BID: Any bid within trigger window extends the auction
 * - L1_CHANGE: Only extends if bid becomes new lowest price
 * - RANK_CHANGE: Extends if bid changes supplier rankings
 *
 * Returns an object with the new close time and string reason, or null.
 */
export async function evaluateExtension(
  tx: any,
  rfqId: number,
  supplierId: number,
  bidPrice: number,
  bidTimestamp: Date
): Promise<{ newClose: Date, reason: string } | null> {
  const rfq = await tx.rfq.findUnique({ where: { id: rfqId } });
  if (!rfq) return null;

  // Reject bids outside the active window.
  if (bidTimestamp < rfq.start_time || bidTimestamp > rfq.close_time) {
    throw { status: 400, message: 'Auction is not currently active' };
  }

  const triggerWindowMs = rfq.trigger_window_mins * 60 * 1000;
  const isWithinTriggerWindow = rfq.close_time.getTime() - bidTimestamp.getTime() <= triggerWindowMs;

  if (!isWithinTriggerWindow) return null;

  let shouldExtend = false;
  let reasonString = '';

  if (rfq.trigger_type === 'ANY_BID') {
    shouldExtend = true;
    reasonString = 'Extended due to ANY_BID rules';
  } else {
    // We need to calculate the current leaderboard state *before* this bid conceptually.
    // Fetch the lowest active bid per supplier for this RFQ.
    const allBids = await tx.bid.groupBy({
      by: ['supplierId'],
      where: { rfqId },
      _min: { price: true },
    });

    const leaderboard = allBids
      .map((b: any) => ({ supplierId: b.supplierId, lowest: b._min.price }))
      .sort((a: any, b: any) => a.lowest - b.lowest);

    if (rfq.trigger_type === 'L1_CHANGE') {
      const currentL1 = leaderboard.length > 0 ? leaderboard[0] : null;
      if (!currentL1 || bidPrice < currentL1.lowest) {
        shouldExtend = true;
        reasonString = 'Extended due to L1_CHANGE: Supplier claimed 1st place rank';
      }
    } else if (rfq.trigger_type === 'RANK_CHANGE') {
      const currentRankIndex = leaderboard.findIndex((s: any) => s.supplierId === supplierId);
      
      // If they had no previous rank, and there are people on the board, they might displace someone immediately.
      // But let's build the "hypothetical" next leaderboard
      const hypotheticalBoard = [...leaderboard];
      if (currentRankIndex > -1) {
        // Only updates their own score if they improved it
        if (bidPrice < hypotheticalBoard[currentRankIndex].lowest) {
           hypotheticalBoard[currentRankIndex].lowest = bidPrice;
        }
      } else {
        hypotheticalBoard.push({ supplierId, lowest: bidPrice });
      }

      hypotheticalBoard.sort((a: any, b: any) => a.lowest - b.lowest);
      const newRankIndex = hypotheticalBoard.findIndex((s: any) => s.supplierId === supplierId);

      // If they were not on the board before, and now they placed above SOMEONE
      if (currentRankIndex === -1 && newRankIndex < hypotheticalBoard.length - 1) {
        shouldExtend = true;
      } else if (currentRankIndex > -1 && newRankIndex < currentRankIndex) {
        // They moved up in rank
        shouldExtend = true;
      }
      
      if (shouldExtend) {
         reasonString = `Extended due to RANK_CHANGE: Ranked moved up to position ${newRankIndex + 1}`;
      }
    }
  }

  if (shouldExtend) {
    const extMs = rfq.extension_mins * 60 * 1000;
    let newClose = new Date(rfq.close_time.getTime() + extMs);

    // Strict cap by forced close time.
    if (newClose > rfq.forced_close_time) {
      newClose = rfq.forced_close_time;
    }

    if (newClose > rfq.close_time) {
      return { newClose, reason: reasonString };
    }
  }

  return null;
}

File: backend-go/internal/service/auction_service.go
package service

import (
	"british-auction-backend/internal/domain"
	"time"
)

type AuctionService interface {
	EvaluateExtension(rfq *domain.Rfq, newBid *domain.Bid, currentLeaderboard []domain.Bid) *ExtensionResult
}

type ExtensionResult struct {
	ExtendedMins int
	NewCloseTime time.Time
	Reason       string
}

type auctionService struct{}

func NewAuctionService() AuctionService {
	return &auctionService{}
}

func (s *auctionService) EvaluateExtension(rfq *domain.Rfq, newBid *domain.Bid, currentLeaderboard []domain.Bid) *ExtensionResult {
	now := time.Now()
	
	// If outside the trigger window, no extension
	windowStart := rfq.CloseTime.Add(-time.Duration(rfq.TriggerWindowMins) * time.Minute)
	if now.Before(windowStart) {
		return nil
	}

	var shouldExtend bool
	var reason string

	switch rfq.TriggerType {
	case domain.TriggerTypeAnyBid:
		shouldExtend = true
		reason = "Extended due to Any Bid rule within the trigger window"
	case domain.TriggerTypeL1Change:
		if len(currentLeaderboard) == 0 {
			shouldExtend = true
			reason = "Extended due to L1 Change (First bid placed)"
		} else {
			l1Price := currentLeaderboard[0].Price
			if newBid.Price < l1Price {
				shouldExtend = true
				reason = "Extended due to L1 Change (New lowest bid)"
			}
		}
	case domain.TriggerTypeRankChange:
		shouldExtend, reason = s.evaluateRankChange(newBid, currentLeaderboard)
	}

	if !shouldExtend {
		return nil
	}

	// Calculate new close time
	newClose := rfq.CloseTime.Add(time.Duration(rfq.ExtensionMins) * time.Minute)

	// Enforce forced close time (Drop-dead)
	if newClose.After(rfq.ForcedCloseTime) {
		newClose = rfq.ForcedCloseTime
	}

	// If the new close time is not actually later than the current close time, don't extend
	if !newClose.After(rfq.CloseTime) {
		return nil
	}

	extendedMins := int(newClose.Sub(rfq.CloseTime).Minutes())

	return &ExtensionResult{
		ExtendedMins: extendedMins,
		NewCloseTime: newClose,
		Reason:       reason,
	}
}

func (s *auctionService) evaluateRankChange(newBid *domain.Bid, currentLeaderboard []domain.Bid) (bool, string) {
	if len(currentLeaderboard) == 0 {
		return true, "Extended due to Rank Change (First participant)"
	}

	// Find the supplier's current rank (if they have one)
	currentRank := -1
	for i, b := range currentLeaderboard {
		if b.SupplierID == newBid.SupplierID {
			currentRank = i
			break
		}
	}

	// Build a hypothetical new leaderboard
	var newLeaderboard []domain.Bid
	added := false
	for _, b := range currentLeaderboard {
		if b.SupplierID == newBid.SupplierID {
			continue // We'll add the new bid instead
		}
		if !added && newBid.Price < b.Price {
			newLeaderboard = append(newLeaderboard, *newBid)
			added = true
		}
		newLeaderboard = append(newLeaderboard, b)
	}
	if !added {
		newLeaderboard = append(newLeaderboard, *newBid)
	}

	// Find the supplier's new rank
	newRank := -1
	for i, b := range newLeaderboard {
		if b.SupplierID == newBid.SupplierID {
			newRank = i
			break
		}
	}

	// Extension rule:
	// 1. Supplier improved their rank
	// 2. Or supplier is a new entrant who didn't just place last
	if currentRank == -1 {
		if newRank < len(newLeaderboard)-1 {
			return true, "Extended due to Rank Change (New supplier entered and ranked above others)"
		}
	} else if newRank < currentRank {
		return true, "Extended due to Rank Change (Supplier improved rank position)"
	}

	return false, ""
}
